tensorflow/contrib/eager/python/examples/generative_examples/image_captioning_with_attention.ipynb

```diff
@@ -0,0 +1,1188 @@
+        "id": "-bYN7xA0RPGl",
+        "colab_type": "code",
+        "colab": {
+          "autoexec": {
+            "startup": false,
+            "wait_interval": 0
+          }
+        }
+      },
+      "cell_type": "code",
+      "source": [
+        "optimizer = tf.train.AdamOptimizer()\n",
+        "\n",
+        "# We are masking the loss calculated for padding\n",
+        "def loss_function(real, pred):\n",
+        "    mask = 1 - np.equal(real, 0)\n",
+        "    loss_ = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=real, logits=pred) * mask\n",
+        "    return tf.reduce_mean(loss_)"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "PHod7t72RPGn",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "## Training\n",
+        "\n",
+        "* We extract the features stored in the respective `.npy` files and then pass those features through the encoder.\n",
+        "* The encoder output, hidden state(initialized to 0) and the decoder input (which is the start token) is passed to the decoder.\n",
+        "* The decoder returns the predictions and the decoder hidden state.\n",
+        "* The decoder hidden state is then passed back into the model and the predictions are used to calculate the loss.\n",
+        "* Use teacher forcing to decide the next input to the decoder.\n",
+        "* Teacher forcing is the technique where the target word is passed as the next input to the decoder.\n",
+        "* The final step is to calculate the gradients and apply it to the optimizer and backpropagate.\n"
+      ]
+    },
+    {
+      "metadata": {
+        "id": "Vt4WZ5mhJE-E",
+        "colab_type": "code",
+        "colab": {
+          "autoexec": {
+            "startup": false,
+            "wait_interval": 0
+          }
+        }
+      },
+      "cell_type": "code",
+      "source": [
+        "# adding this in a separate cell because if you run the training cell \n",
+        "# many times, the loss_plot array will be reset\n",
+        "loss_plot = []"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "UlA4VIQpRPGo",
+        "colab_type": "code",
+        "colab": {
+          "autoexec": {
+            "startup": false,
+            "wait_interval": 0
+          }
+        }
+      },
+      "cell_type": "code",
+      "source": [
+        "EPOCHS = 20\n",
+        "\n",
+        "for epoch in range(EPOCHS):\n",
+        "    start = time.time()\n",
+        "    total_loss = 0\n",
+        "    \n",
+        "    for (batch, (img_tensor, target)) in enumerate(dataset):\n",
+        "        loss = 0\n",
+        "        \n",
+        "        # initializing the hidden state for each batch\n",
+        "        # because the captions are not related from image to image\n",
+        "        hidden = decoder.initialize_hidden_state()\n",
```

Comment:
This would be easier to follow if the hidden-state was .... hidden.

can you pull [my update](https://github.com/MarkDaoust/tensorflow/commits/image_captioning/tensorflow/contrib/eager/python/examples/generative_examples)? (The web interface won't let me send you a PR)

My change makes the hidden state an attribute of the decoder, so it doesn't need to be managed out here in this loop. It also makes the `reset_states` take a batch size so we don't need the `batch_and_drop_remainder`, can just use batch.

```
git pull git://github.com/MarkDaoust/tensorflow.git image_captioning
```

WDYT? If you take the patch please also apply it to the NMT tutorial.

Reply:
I don't prefer doing this, because it becomes difficult for users to see how the hidden state is being manipulated. I did this in the text generation example and it took me some time to understand what's going on. I can do the batch changes.. its easy to do that. But I would prefer if the hidden state is not an attribute of the decoder. WDYT?

Reply:
> I don't prefer doing this, because it becomes difficult for users to see how the hidden state is being manipulated.

It just seems to me that the state of the object should be in the object. I feel like it's easier to follow this way, because then you only have to read the decoder's code to know everything that happens to that state. 

(Keras RNN layers support this pattern with "stateful=True" but this didn't seem to play nice with the eager setup here.) 

Anyone else have an opinion?

> I can do the batch changes

+1, regardless of the state question above, this seems like a clear win either way.

Reply:
I just think that hidden state being outside, shows that it is being passes back to the model again. Its easier for me to follow. @alextp WDYT?

Reply:
I made the batch change. Waiting on alex when he returns to break this tie